lib/Sanger/CGP/NgsQc/ValidateSampleMetaAndBamHeader.pm

```diff
@@ -0,0 +1,755 @@
+        $donor_uuid = $exist_donor_uuid{$donor_ids[$i]};
+      }
+
+      if (! exists $exist_tissue_uuid{$tissue_ids[$i]}) {
+        warn "generating tissue UUID for: $tissue_ids[$i]\n";
+        $tissue_uuid = get_uuid();
+        $exist_tissue_uuid{$tissue_ids[$i]} = $tissue_uuid;
+      } else {
+        warn "use existing tissue UUID for: $tissue_ids[$i]\n";
+        $tissue_uuid = $exist_tissue_uuid{$tissue_ids[$i]};
+      }
+      warn "generating sample UUID for: $sample_ids[$i]\n";
+      $sample_uuid = get_uuid();
+
+      my $left = scalar @bam_files - $i - 1;
+      warn "generating md5sum for: $bam_files[$i]. [$left bam(s) to go]\n";
+      $md5sum = get_md5sum($bam_files[$i]);
+      push @out, $lines[$i]."\t".$bam_states[$i]."\t".$donor_uuid."\t".$tissue_uuid."\t".$sample_uuid."\t".$md5sum;
+      warn "done!\n--------\n";
+    }
+  }
+
+  if ($opts->{'mod'} == 1) {
+    write_results(\@out, $opts->{'o'}, $opts->{'f'}, $traffic_light);
+    return ($traffic_light);
+  } else { # return array ref when in test mode, instead of writing to file
+    return (\@out, $traffic_light);
+  }
+}
+
+sub get_md5sum {
+  my $file = shift;
+  open(FILE, "<", $file);
+  my $cksum = md5_hex(<FILE>);
+  if ($cksum) {
+    return $cksum;
+  } else {
+    return 0;
+  }
+}
+
+sub get_uuid {
+  my $ug = Data::UUID->new;
+  my $uuid = $ug->to_string($ug->create);
+  return $uuid;
+}
+
+sub validate_bam {
+  my ($sample_ID, $is_normals, $align_file, $check_all, $count_base_number) = @_; # $count_base_number is for testing the script
+  my $bam_state = 'failed:';
+
+  if ($is_normals !~ m/^yes$|^y$|^no$|^n$/i) {
+    $bam_state .= $FAIL_CODE_010."; ";
+  }
+
+  if (! -e $align_file) {
+    warn "can not find the file: $align_file\n";
+    return $bam_state.$FAIL_CODE_020.";";
+  }
+
+  warn "--------\nstart to validate alignment file: $align_file\n--------\n";
+  my $sam = Bio::DB::HTS->new(-bam => $align_file);
+  my @lines = split "\n", $sam->header->text;
+  my %headers; # to handle duplicates!
+
+  my (@RG_lines, @SQ_lines, @PG_lines);
+  for(@lines) {
+    push @RG_lines, $_ if(index($_,'@RG') == 0);
+    push @SQ_lines, $_ if(index($_,'@SQ') == 0);
+    push @PG_lines, $_ if(index($_,'@PG') == 0);
+  }
+
+  if(scalar @RG_lines != 0) {
+    my %reported_error;
+    for (@RG_lines) {
+      my ($id) = "$_\t" =~ m/ID:([^\t]+)/;
+      if (defined $id) {
+        if (! exists $headers{'RG'}{$id} ) {
+          $headers{'RG'}{$id} = 0; # handle duplicates also = 1 when rg found in a read.
+          my ($pl) = "$_\t" =~ m/PL:([^\t]+)/;
+          if (! defined $pl && ! exists $reported_error{$FAIL_CODE_070}) {
+            warn "no PL for RG $id.\n";
+            $reported_error{$FAIL_CODE_070} = 1;
+            $bam_state .= $FAIL_CODE_070."; ";
+          } else {
+            if ($pl !~ m/^CAPILLARY$|^ILLUMINA$|^LS454$|^SOLID$|^HELICOS$|^IONTORRENT$|^ONT$|^PACBIO$/ && ! exists $reported_error{$FAIL_CODE_080}) {
```

Comment:
Use a constant array with 'first', easier to modify later and makes the block of code self documenting

```perl
use List::Util qw(first);
const my @VALID_PLATFORM => qw (CAPILLARY ILLUMINA LS454 SOLID HELICOS IONTORRENT ONT PACBIO);
...
unless(first { $_ eq $pl } @VALID_PLATFORM) {
  ...
}
```

Oddly `first` is faster than `any` (at least it was the last time I did a benchmark).